File: scala/bob/src/main/scala/Bob.scala
object Bob {

  def response(statement: String): String =
    if (isShouting(statement) && isQuestion(statement))
      "Calm down, I know what I'm doing!"
    else if (isShouting(statement))
      "Whoa, chill out!"
    else if (isQuestion(statement))
      "Sure."
    else if (isEmpty(statement))
      "Fine. Be that way!"
    else
      "Whatever."

  def isQuestion(statement: String): Boolean =
    statement.reverse.dropWhile(_.isWhitespace).startsWith("?")

  def isShouting(statement: String): Boolean =
    !statement.exists(_.isLower) && statement.exists(_.isUpper)

  def isEmpty(statement: String): Boolean = statement.forall(_.isWhitespace)
}

File: clojure/bob/src/bob.clj
(ns bob)

(defn lower? [s]
  (re-find #"[a-z]" s))

(defn upper? [s]
  (re-find #"[A-Z]" s))

(defn shout? [s]
  (and (upper? s) (not (lower? s))))

(defn question? [s]
  (re-matches #".*\?$" s))

(defn nothing? [s]
  (re-matches #"\W*" s))

(defn response-for [s]
  (cond
    (and (shout? s) (question? s)) "Calm down, I know what I'm doing!"
    (shout? s) "Whoa, chill out!"
    (question? s) "Sure."
    (nothing? s) "Fine. Be that way!"
    :else "Whatever."))
